_posts/2021-02-24-#21224.md

```diff
@@ -0,0 +1,79 @@
+authors: [ariard]
+components: ["p2p"]
+host: ellemouton
+status: upcoming
+commit:
+---
+
+## Notes
+
+- Transaction relay is a three step process: `inv` -> `getdata` -> `tx`:
+
+  1. the relaying node sends an `inv` message to the receiving node to announce
+     a new transaction.
+  2. the receiving node sends a `getdata` message to the relaying node to
+     request the full transaction.
+  3. the relaying node delivers a `tx` message to the receiving node. If the
+     relaying node is no longer able to deliver the transaction, it responds
+     with `notfound` instead of the `tx`.
+
+  You can learn more about these messages [here](https://en.bitcoin.it/wiki/Protocol_documentation#inv).
+
+- Currently if a `tx` message is received it will not be processed if:
+
+	1. The node is in blocks-only mode and the peer sending the message has 
+	no relay permissions
+	2. The peer is a blocks-relay only peer
+
+  In all other cases, the node will process a transaction that it receives even 
+  if it has not made a request for that transaction (i.e: If it recieves a `tx` 
+  message without having first sent a `getdata` message asking for the `tx` message 
+  then it will still process the transaction). In other words the 3 step process
+  described above is not enforced.
+
+- This PR aims to change that by requiring a transaction to first be requested
+  by a node before it processes an accociated `tx` message.
+
+- See the PR’s author's motivation behind the PR along with some of the
+  concerns that this change brings about:
+  [see mailing list thread here](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018391.html)
+
+## Questions
+
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+   NACK?](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)
+
+2. What is the `TxRequestTracker` class used for? See [`src/txrequest.h`](https://github.com/bitcoin/bitcoin/blob/b805dbb0b9c90dadef0424e5b3bf86ac308e103e/src/txrequest.h)
+
+3. What are the different states that a peer/tx “announcement” can be in? Where
+   in the code is an announcement shifted between the different states?
+
+4. What does the author claim that the problem is with the current way in which 
+   transaction messages are handled?
+
+5. What is the change that this PR is making and how does it solve the problem
+   mentioned in Q4?
+
+6. There are a lot of switches on the `PR_RELAY` flag in code relevant to this
+   PR. What does the `PF_RELAY` flag mean? In what case(s) would you want to set
+   this flag for a peer?
+
+7. Discussion: Since the `inv` -> `getdata` -> `tx` sequence has not been necessary
+   for communicating and receiving transaction data, some other clients don't
+   bother with the sequence at all. This means that  if this change was deployed
+   to Bitcoin Core nodes then other clients would not be able to relay transactions 
+   and so they will have to adapt and update. Do the pros out weigh these
```

Comment:
"not be able to relay transactions to upgraded Bitcoin Core nodes. Eventually upgraded nodes would make up the majority of the network, and so those clients would have to ..."